### ISSUE TITLE
Prevent `at-rule-no-unknown` from warning against Sass directives

### ISSUE THREAD

**Issue by VincentSmedinga** [stylelint/stylelint#1612]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

A feature request to not warn against Sass directives like `@import`, `@function`, and `@if` when using Sass or SCSS syntax.

> Which rule, if any, is this issue related to?

`at-rule-no-unknown`

 > What CSS is needed to reproduce this issue?

 e.g.

```scss
@function foo ($bar) {
  @if ($bar == 'a') {
    $baz: 'b';
  }
  @else if ($bar == 'b') {
    $baz: 'a';
  }
  @return $baz;
}
```

> What stylelint configuration is needed to reproduce this issue?

e.g.

```json
{
  "rules": {
    "at-rule-no-unknown": true,
  }
}
```

> Which version of stylelint are you using?

`6.9.0`.

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI with `stylelint foo.scss --config .stylelintrc --syntax scss`.

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

Yes, it’s related to SCSS directives.

> What did you expect to happen?

No warnings to be flagged.

> What actually happened (e.g. what warnings or errors you are getting)?

The following warnings were flagged:

```shell
foo.scss
1:1  ✖  Unexpected unknown at-rule "@function"   at-rule-no-unknown
2:3  ✖  Unexpected unknown at-rule "@if"         at-rule-no-unknown
5:3  ✖  Unexpected unknown at-rule "@else"       at-rule-no-unknown
8:3  ✖  Unexpected unknown at-rule "@return"     at-rule-no-unknown
```

Going over the [Sass documentation](http://sass-lang.com/documentation/file.SASS_REFERENCE.html#control_directives__expressions) results in the following list of keywords to additionally accept as known at-rules: `@each`, `@else`, `@extend`, `@for`, `@function`, `@if`, `@include`, `@mixin`, `@return`, `@while`.

The simplest solution would be to add these to the `atRules` set in `src/reference/keywordSets.js`, but that would incorrectly accept them in plain CSS. Creating a separate set for Sass directives will prevent that, but I can’t assess whether it’s easy to only consider that set in particular syntaxes.

It would be great if stylelint automatically applied this extra set when in Scss or Sass syntax.

**Comment by evilebottnawi** [comment]
@VincentSmedinga `stylelint` rules not check non-standard syntax(not handling, not ignore and etc), use option `ignoreAtRules: ['function', 'if']` for resolve this issue. We offically support only css syntax in rules. Maybe you are interested in `stylelint-scss`(some rules for only scss).

**Comment by VincentSmedinga** [comment]
Ah ok, thanks.

**Comment by jeddy3** [comment]
@VincentSmedinga Thanks for detailed write-up.

As @evilebottnawi said, this is unlikely to happen in core. As a way of an explanation, there is a [ever-increasing and rapidly changing amount](https://github.com/postcss/postcss/blob/master/docs/plugins.md#language-extensions) of at-rule based processor language extensions. We have to draw the line somewhere, and we decided that the most sustainable line would be between standard CSS and processor CSS. 

To assist users we've added `ignore*: []` secondary options to most rules, like the `ignoreAtRules: ['function', 'if']` that @evilebottnawi mentioned. The gaps created by these secondary options can then be filled by community plugins (packs) like [stylelint-scss](https://github.com/kristerkari/stylelint-scss).

I'm going to reopen the issue though and mark it as "documentation" as I think we can better communicate in the READMEs that the `*-no-unknown` rules are only applicable to standard-css. I suspect this issue will be raised again if we don't.

**Comment by VincentSmedinga** [comment]
@jeddy3 I hadn’t noticed your further reply until now, thanks for clarifying.

**Comment by benbonnet** [comment]
sorry to bother here; @mixin / @include sounds common. google around, the issue is know, but the solution is far from being explicit anywhere

How one would turn that one error off ?

**Comment by hudochenkov** [comment]
@benbonnet [How do I turn off, disable or ignore a rule?](https://stylelint.io/user-guide/faq/#how-do-i-turn-off-disable-or-ignore-a-rule)

Also, https://github.com/stylelint/stylelint/issues/3190#issuecomment-368350317

**Comment by benbonnet** [comment]
thx; sorry for being a noob there; I guess i've been a bit confused by the [explanations here](https://stylelint.io/user-guide/rules/at-rule-no-unknown/) :

```
ignoreAtRules: ["/regex/", "string"]
```

Could not make it work correctly

It worked out perfectly  with the following in .stylelintrc :

```
"rules": {
   ...
    "at-rule-no-unknown": null
 }
```

**Comment by datatypevoid** [comment]
@benbonnet 
```
"rules": {
    . . .
    "at-rule-no-unknown": [true, {
      "ignoreAtRules": ["function", "if", "each", "include", "mixin"]
    }]
    . . .
  }

Reference: https://stylelint.io/user-guide/configuration/#the-configuration-object
```

**Comment by cooperka** [comment]
Clearly this issue is still popular -- I'm now using https://github.com/bjankord/stylelint-config-sass-guidelines for automatically setting rules like this and it works great for me.

```json
{
  "extends": "stylelint-config-sass-guidelines"
}
```

**Comment by gimler** [comment]
install https://github.com/kristerkari/stylelint-scss and then see https://github.com/kristerkari/stylelint-scss/tree/master/src/rules/at-rule-no-unknown
```
"rules": {
    "at-rule-no-unknown": null,
    "scss/at-rule-no-unknown": true
  }
```
